### ISSUE TITLE
Need some clarity in the training stage on custom new language data

### ISSUE THREAD

**Issue by kafan1986** [k2-fsa/icefall#1031]
I am training the Zipformer on custom language dataset. The training data is already prepared according the librispeech format i.e. each line contains the audiofile name followed by a single space and then the ground truth transcript. I have few doubts.

1) Is the language model a necessity during training phase?
2) How to train a new language model for my new language, kenLM (ngram based one) as zipformer is using BPE based tokenizer. Currently in the example script the English language model are downloaded from openslr CDN

**Comment by kafan1986** [comment]
> 1. You don't need a language model when training a zipformer transducer.

And during later inference phase, do I need language model, say for greedy decoding? Anyways, would like to have the instructions for training the n-gram based LM for increased accuracy.

**Comment by csukuangfj** [comment]
> > 1. You don't need a language model when training a zipformer transducer.
> 
> 
> 
> Are you sure about this, the code base looks like to be downloading the LM model automatically 
> 
> 
> 
> https://github.com/k2-fsa/icefall/blob/61ec3a7a8fc8be859b23a821e568950fb898b37a/egs/librispeech/ASR/prepare.sh#LL69C11-L69C11
> 
> 
> 
> As my language is a different from regular english, I want to prevent this.
> 
> 
> 
> And during later inference phase, do I need language model, say for greedy decoding? Anyways, would like to have the instructions for training the n-gram based LM for increased accuracy. 

Yes, I am pretty sure.

prepare.sh is shared by various recipes, e.g., CTC and transducer.

The downloaded LM can be used for CTC decoding or transducer decoding, but you don't need an LM during transducer training.

For greedy search, modified beam search and fast beam search of transducer models, you also don't need an LM. 

By the way, n-gram LM is not limited to words, you can also train a BPE-based n-gram LM. Please check prepare.sh from librispeech for how to do that.
